File: plugins/plugin-elizacloud/python/elizaos_plugin_elizacloud/models/object.py
import json
import re

from elizaos_plugin_elizacloud.providers import ElizaCloudClient
from elizaos_plugin_elizacloud.types import ElizaCloudConfig, ObjectGenerationParams


def _parse_json_response(content: str) -> dict[str, object]:
    json_match = re.search(r"```(?:json)?\s*([\s\S]*?)```", content)
    if json_match:
        content = json_match.group(1).strip()

    try:
        return dict(json.loads(content))
    except json.JSONDecodeError:
        obj_match = re.search(r"\{[\s\S]*\}", content)
        if obj_match:
            return dict(json.loads(obj_match.group(0)))
        raise


async def handle_object_small(
    config: ElizaCloudConfig,
    params: ObjectGenerationParams,
) -> dict[str, object]:
    from elizaos_plugin_elizacloud.types import TextGenerationParams

    enhanced_prompt = f"{params.prompt}\n\nRespond with valid JSON only."

    async with ElizaCloudClient(config) as client:
        text = await client.generate_text(
            TextGenerationParams(
                prompt=enhanced_prompt,
                temperature=params.temperature,
            ),
            model_size="small",
        )
        return _parse_json_response(text)


async def handle_object_large(
    config: ElizaCloudConfig,
    params: ObjectGenerationParams,
) -> dict[str, object]:
    from elizaos_plugin_elizacloud.types import TextGenerationParams

    enhanced_prompt = f"{params.prompt}\n\nRespond with valid JSON only."

    async with ElizaCloudClient(config) as client:
        text = await client.generate_text(
            TextGenerationParams(
                prompt=enhanced_prompt,
                temperature=params.temperature,
            ),
            model_size="large",
        )
        return _parse_json_response(text)

File: plugins/plugin-elizacloud/rust/src/models/object.rs
#![allow(missing_docs)]

use crate::error::Result;
use crate::providers::ElizaCloudClient;
use crate::types::{ElizaCloudConfig, ObjectGenerationParams, TextGenerationParams};

fn parse_json_response(content: &str) -> Result<serde_json::Value> {
    let content = if let Some(start) = content.find("```") {
        let after_backticks = &content[start + 3..];
        let json_start = after_backticks.find('\n').map(|i| i + 1).unwrap_or(0);
        let after_lang = &after_backticks[json_start..];
        if let Some(end) = after_lang.find("```") {
            after_lang[..end].trim()
        } else {
            content.trim()
        }
    } else {
        content.trim()
    };

    match serde_json::from_str(content) {
        Ok(v) => Ok(v),
        Err(_) => {
            if let Some(start) = content.find('{') {
                if let Some(end) = content.rfind('}') {
                    let json_str = &content[start..=end];
                    serde_json::from_str(json_str).map_err(Into::into)
                } else {
                    serde_json::from_str(content).map_err(Into::into)
                }
            } else {
                serde_json::from_str(content).map_err(Into::into)
            }
        }
    }
}

pub async fn handle_object_small(
    config: ElizaCloudConfig,
    params: ObjectGenerationParams,
) -> Result<serde_json::Value> {
    let client = ElizaCloudClient::new(config)?;

    let enhanced_prompt = format!("{}\n\nRespond with valid JSON only.", params.prompt);

    let text = client
        .generate_text_small(TextGenerationParams {
            prompt: enhanced_prompt,
            temperature: params.temperature,
            ..Default::default()
        })
        .await?;

    parse_json_response(&text)
}

pub async fn handle_object_large(
    config: ElizaCloudConfig,
    params: ObjectGenerationParams,
) -> Result<serde_json::Value> {
    let client = ElizaCloudClient::new(config)?;

    let enhanced_prompt = format!("{}\n\nRespond with valid JSON only.", params.prompt);

    let text = client
        .generate_text_large(TextGenerationParams {
            prompt: enhanced_prompt,
            temperature: params.temperature,
            ..Default::default()
        })
        .await?;

    parse_json_response(&text)
}
